workers/unity/Packages/com.improbable.gdk.transformsynchronization/TransformUtils.cs

```diff
@@ -11,51 +11,146 @@ public static bool HasChanged(Coordinates a, Coordinates b)
+                    uint magnitude = (uint) (((1u << 9) - 1u) * (Mathf.Abs(q[i]) / SqrtHalf) + 0.5f);
+                    compressedQuaternion = (compressedQuaternion << 10) | (signBit << 9) | magnitude;
+                }
+            }
+
+            return new CompressedQuaternion(compressedQuaternion);
         }
 
-        public static UnityEngine.Quaternion ToUnityQuaternion(this Quaternion quaternion)
+        public static UnityEngine.Vector3 ToUnityVector3(this FixedPointVector3 fixedPointVector3)
         {
-            return new UnityEngine.Quaternion(quaternion.X, quaternion.Y, quaternion.Z, quaternion.W);
+            return new Vector3
+            {
+                x = FixedToFloat(fixedPointVector3.X),
+                y = FixedToFloat(fixedPointVector3.Y),
+                z = FixedToFloat(fixedPointVector3.Z)
+            };
         }
 
-        public static Quaternion ToImprobableQuaternion(this UnityEngine.Quaternion quaternion)
+        public static FixedPointVector3 ToFixedPointVector3(this Vector3 vector3)
         {
-            return new Quaternion(quaternion.w, quaternion.x, quaternion.y,
-                quaternion.z);
+            return new FixedPointVector3
+            {
+                X = FloatToFixed(vector3.x),
+                Y = FloatToFixed(vector3.y),
+                Z = FloatToFixed(vector3.z)
+            };
         }
 
-        public static UnityEngine.Vector3 ToUnityVector3(this Location location)
+        public static FixedPointVector3 ToFixedPointVector3(this Coordinates coordinates)
         {
-            return new Vector3(location.X, location.Y, location.Z);
+            return new FixedPointVector3
+            {
+                X = FloatToFixed((float) coordinates.X),
+                Y = FloatToFixed((float) coordinates.Y),
+                Z = FloatToFixed((float) coordinates.Z)
+            };
         }
 
-        public static Location ToImprobableLocation(this Vector3 vector)
+        public static Coordinates ToCoordinates(this Vector3 vector3)
         {
-            return new Location(vector.x, vector.y, vector.z);
+            return new Coordinates
+            {
+                X = vector3.x,
+                Y = vector3.y,
+                Z = vector3.z
+            };
         }
 
-        public static UnityEngine.Vector3 ToUnityVector3(this Velocity velocity)
+        public static Coordinates ToCoordinates(this FixedPointVector3 fixedPointVector3)
         {
-            return new Vector3(velocity.X, velocity.Y, velocity.Z);
+            return new Coordinates
+            {
+                X = FixedToFloat(fixedPointVector3.X),
+                Y = FixedToFloat(fixedPointVector3.Y),
+                Z = FixedToFloat(fixedPointVector3.Z)
+            };
         }
 
-        public static Velocity ToImprobableVelocity(this Vector3 velocity)
+        private const int FixedPointOne = 0x00010000;
```

Comment:
Is this accuracy deliberate? I would have expected you to want 10 bits (mm accuracy) after the binary point and 21 bits before + a sign bit.
This looks like you are going down to about half a micrometer accuracy, which would give you about 16km to play with if you are feeding this from a double. But from a float you can only represent ~ 2^24 as an integer without loss of precision. Which means that you can go up to about +-256m before it loses precision. Arguably this loss in prevision isn't important as it will be in the bit that probably already has too much precision, but it still seems a bit odd. It might be worth working out what precision loss is expected at expected values.

I'd also not do this is a single hex string. Make it something like `1u << 10` to make the precision clear. Also probably put a comment on what rough decimal range this corresponds too.

Reply:
It's not deliberate - you're right that we probably we don't need more than mm accuracy